js/ui/components/clubhouse.js

```diff
@@ -112,6 +112,148 @@ var ClubhouseNotificationSource = new Lang.Class({
+
+        this._closeButton =
+            new St.Button({ name: 'clubhouse-close-button',
+                            child: new St.Icon({ style_class: 'clubhouse-close-button-icon' }),
+                          });
+        this._closeButton.connect('clicked', () => {
+            this.emit('close-clubhouse');
+        })
+
+        Main.layoutManager.addChrome(this._closeButton);
+
+        this._clubhouseWindowActor = null;
+        this._trackWindowActor(this._getClubhouseActorFromWM());
+
+        this.updateVisibility();
+        this.reposition();
+
+        this._windowMapHandler =
+            global.window_manager.connect('map', Lang.bind(this, this._mapWindow));
+
+        this._overviewShowingHandler = Main.overview.connect('showing', () => {
+            this.updateVisibility();
+        });
+
+        this._overviewHiddenHandler = Main.overview.connect('hidden', () => {
+            this.updateVisibility();
+        });
+    },
+
+    _getClubhouseActorFromWM: function() {
+        for (let actor of global.get_window_actors()) {
+            let window = actor.meta_window;
+            if (window && window.get_gtk_application_id() == CLUBHOUSE_ID)
+                return actor;
+        }
+
+        return null;
+    },
+
+    _updateCloseButtonPosition: function() {
+        this._closeButton.y = this._openButton.y;
+        if (this._clubhouseWindowActor)
+            this._closeButton.x = this._clubhouseWindowActor.x - this._iconSize / 2;
+    },
+
+    reposition: function() {
+        let monitor = Main.layoutManager.primaryMonitor;
+        if (!monitor)
+            return;
+
+        this._openButton.x = monitor.width - this._iconSize / 2;
+        this._openButton.y = Math.floor(monitor.height / 2.0 - this._iconSize / 2.0);
+
+        this._updateCloseButtonPosition();
+    },
+
+    updateVisibility: function() {
+        this._closeButton.visible = !Main.overview.visible && this._clubhouseWindowActor;
+        this._openButton.visible = !this._closeButton.visible;
+    },
+
+    _mapWindow: function(shellwm, actor) {
+        let metaWindow = actor.meta_window;
+        if (metaWindow && metaWindow.get_gtk_application_id() == CLUBHOUSE_ID)
+            this._trackWindowActor(actor);
+    },
+
+    _trackWindowActor: function(actor) {
+        if (this._clubhouseWindowActor == actor)
+            return;
+
+        // Reset the current window actor
+        if (this._clubhouseWindowActor) {
+            this._clubhouseWindowActor.disconnect(this._clubhouseNotifyHandler);
+            this._clubhouseNotifyHandler = 0;
+
+            this._clubhouseWindowActor.disconnect(this._clubhouseDestroyHandler);
+            this._clubhouseDestroyHandler = 0;
+        }
+
+        this._clubhouseWindowActor = actor;
+
+        if (!actor)
+            return;
+
+        // Track new window actor
+        this._clubhouseNotifyHandler = actor.connect('notify::x', (actor, prop) => {
```

Comment:
I am not sure why you need to keep track of the `x` property to determine what feels like the visibility of the clubhouse window. Don't we already have a visible property in the DBus interface that can be used for this?

Reply:
It's not for the visibility, it's for setting the `closeButton`'s position. I may have left those visibility bits after one of the many attempts to get it right. I will see if I can remove those bits and leave only the close button's position.